smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/waiters/JMESPathVisitor.kt

```diff
@@ -0,0 +1,412 @@
+                        addTempVar(flattenedVar, "\$L\$LflatMap { $$0 }", toBeFlattened.name, dotOperator)
+                    }
+                    else -> {
+                        // Single nested List.  Return original list unchanged.
+                        toBeFlattened
+                    }
+                }
+            }
+            else -> {
+                toBeFlattened
+            }
+        }
+    }
+
+    // Implement contains() and length() free functions which are the only 2 JMESPath methods we support.
+    // contains() returns true if its 1st param is a collection that contains an element equal
+    // to the 2nd param, false otherwise.
+    // length() returns the number of elements of an array, the number of key/value pairs for a map,
+    // or the number of characters for a string.  Zero is returned if the argument is nil.
+    override fun visitFunction(expression: FunctionExpression): Variable {
+        when (expression.name) {
+            "contains" -> {
+                if (expression.arguments.size != 2) {
+                    throw Exception("Unexpected number of arguments to $expression")
+                }
+                val subject = expression.arguments[0]
+                val subjectVariable = subject.accept(this)
+                val search = expression.arguments[1]
+                val searchVariable = search.accept(this)
+                val optionalityMark = "?".takeIf { subjectVariable.isOptional } ?: ""
+                val returnValueVar = Variable("contains", false, boolShape)
+                return if (searchVariable.isOptional) {
+                    addTempVar(
+                        returnValueVar,
+                        "\$L.flatMap { \$L\$L.contains($$0) } ?? false",
+                        searchVariable.name,
+                        subjectVariable.name,
+                        optionalityMark
+                    )
+                } else {
+                    addTempVar(
+                        returnValueVar,
+                        "\$L\$L.contains(\$L)",
+                        subjectVariable.name,
+                        optionalityMark,
+                        searchVariable.name
+                    )
+                }
+            }
+            "length" -> {
+                if (expression.arguments.size != 1) {
+                    throw Exception("Unexpected number of arguments to $expression")
+                }
+                val subjectExp = expression.arguments[0]
+                val subject = subjectExp.accept(this)
+
+                return when (subject.shape) {
+                    is StringShape, is ListShape, is MapShape -> {
+                        val countVar = Variable("count", false, doubleShape)
+                        val optionalityMark = "?".takeIf { subject.isOptional } ?: ""
+                        addTempVar(countVar, "Double(\$L\$L.count ?? 0)", subject.name, optionalityMark)
```

Comment:
We likely only want to do the nil coalescing if the subject is optional, otherwise this will result in a warning

Reply:
Nil-coalescence here is now added only if the subject is Optional.